### PR TITLE
IGNITE-11898 Java Thin: Implement Best Effort Affinity

### PR DESCRIPTION
**Pull request by alex-plekhanov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheAffinityMapping.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+     * @param key Key.
+     * @return Affinity node id or {@code null} if affinity node can't be determined for given cache and key.
+     */
+    public UUID affinityNode(IgniteBinary binary, int cacheId, Object key) {
+        CacheAffinityInfo affinityInfo = cacheAffinity.get(cacheId);
+
+        if (affinityInfo == null || affinityInfo.keyCfg == null || affinityInfo.partMapping == null)
+            return null;
+
+        Object binaryKey = binary.toBinary(key);
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheAffinityMapping.java]: Check for simple types is already exists inside ClientBinary.toBinary method (there also was an error in type check here and I fixed it with `BinaryUtils.isBinaryType(obj.getClass())`)

[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheAffinityMapping.java]: I see, thanks.

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAffinityAwarenessUnstableTopologyTest.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+package org.apache.ignite.internal.client.thin;
+
+import org.apache.ignite.client.ClientCache;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.junit.Test;
+
+/**
+ * Test affinity awareness of thin client on unstable topology.
+ */
+public class ThinClientAffinityAwarenessUnstableTopologyTest extends ThinClientAbstractAffinityAwarenessTest {
```
Comment (pavlukhin) [inline modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAffinityAwarenessUnstableTopologyTest.java]: Looks like a test was not added to a suite.

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractAffinityAwarenessTest.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+
+import static org.apache.ignite.configuration.ClientConnectorConfiguration.DFLT_PORT;
+
+/**
+ * Abstract thin client affinity awareness test.
+ */
+public class ThinClientAbstractAffinityAwarenessTest extends GridCommonAbstractTest {
```
Comment (pavlukhin) [inline modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractAffinityAwarenessTest.java]: Can we make this class abstract?

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAffinityAwarenessUnstableTopologyTest.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAffinityAwarenessUnstableTopologyTest.java]: Forgot it. Fixed, thank you.

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractAffinityAwarenessTest.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/core/src/test/java/org/apache/ignite/internal/client/thin/ThinClientAbstractAffinityAwarenessTest.java]: Sure, fixed.

**Pull request closed without merging**